h2o-automl/src/main/java/ai/h2o/automl/TrainingStepsRegistry.java

```diff
@@ -0,0 +1,106 @@
+package ai.h2o.automl;
+
+import ai.h2o.automl.EventLogEntry.Stage;
+import ai.h2o.automl.StepDefinition.Alias;
+import ai.h2o.automl.StepDefinition.Step;
+import water.Iced;
+import water.nbhm.NonBlockingHashMap;
+import water.util.ArrayUtils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The registry responsible for loading all {@link TrainingStepsProvider} using service discovery,
+ * and providing the list of {@link TrainingStep} to execute.
+ */
+public class TrainingStepsRegistry extends Iced<TrainingStepsRegistry> {
+
+    static final NonBlockingHashMap<String, Class<TrainingSteps>> stepsByName = new NonBlockingHashMap<>();
+
+    static {
+        ServiceLoader<TrainingStepsProvider> trainingStepsProviders = ServiceLoader.load(TrainingStepsProvider.class);
+        for (TrainingStepsProvider provider : trainingStepsProviders) {
+            stepsByName.put(provider.getName(), provider.getStepsClass());
```

Comment:
is there a reason why the providers are not factories for the training steps - what is the intention of registering a class instead of the provider?

Reply:
Very good point, don't know why they are not factories taking `AutoML` param, probably because I introduced them a posteriori... will try to turn them into proper factories and get rid of the Java reflection.